src/cargo/core/compiler/custom_build.rs

```diff
@@ -385,12 +396,14 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
         state.running(&cmd);
         let timestamp = paths::set_invocation_time(&script_run_dir)?;
         let prefix = format!("[{} {}] ", id.name(), id.version());
-        let mut warnings_in_case_of_panic = Vec::new();
+        let mut diagnostics = Vec::new();
         let output = cmd
             .exec_with_streaming(
                 &mut |stdout| {
                     if let Some(warning) = stdout.strip_prefix(CARGO_WARNING) {
-                        warnings_in_case_of_panic.push(warning.to_owned());
+                        diagnostics.push(Diagnostic::Warning(warning.to_owned()));
+                    } else if let Some(error) = stdout.strip_prefix(CARGO_ERROR) {
+                        diagnostics.push(Diagnostic::Error(error.to_owned()));
```

Comment:
Please add a test for this

Reply:
It is covered by `custom_build_script_failed_custom_error` test

Reply:
That test is in the commit for `+=` and will be split out of this PR.

Reply:
I've moved part of the test to earlier commits